### ISSUE TITLE
20-minute startup delay: check_memory reads all images from HDF5 before caching

### ISSUE THREAD

**Issue by talmo** [talmolab/sleap-nn#486]
## Summary

When training with `data_pipeline_fw: torch_dataset_cache_img_memory` on a 10K-image `.pkg.slp` file, there is a **~20-minute silent delay** between Trainer initialization and the start of in-memory caching. This is caused by `check_memory()` reading every image from the HDF5 file just to estimate memory requirements — and then the images are discarded and re-read during the actual cache fill.

## Observed behavior

Training logs show a 21-minute gap (01:19:10 → 01:40:36) with no output:

```
2026-03-12 01:19:10 | Trainer devices: 1
2026-03-12 01:19:10 | Using GPUs with most available memory: [0]
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
💡 Tip: ...
2026-03-12 01:40:36 | Using in-memory caching with 4 DataLoader workers. Estimated additional memory for workers: 12.30 GB
2026-03-12 01:40:37 | Caching 10000 images to memory...
2026-03-12 01:49:23 | Caching complete.
2026-03-12 01:49:23 | Caching 2000 images to memory...
2026-03-12 01:50:05 | Caching complete.
```

Note that the actual caching of 10K images only takes ~9 minutes (with `parallel_caching: true` and `cache_workers: 16`), while the silent memory estimation step takes ~21 minutes.

## Root cause trace

The call chain during the gap:

1. `model_trainer.py:1206` — `self._setup_datasets()` is called after `L.Trainer(...)` creation
2. `model_trainer.py:816` — Since `data_pipeline_fw == "torch_dataset_cache_img_memory"`, calls `check_cache_memory()`
3. `utils.py:250` → `estimate_cache_memory()` → `check_memory()` for both train and val labels
4. **`utils.py:95-117` — `check_memory()` iterates over all 12K labeled frames, accessing `label.image` on each one**

```python
def check_memory(labels: sio.Labels) -> int:
    imgs_bytes = []
    for label in labels:
        if label.image is not None:
            img = label.image
            img_bytes = img.nbytes
            imgs_bytes.append(img_bytes)
        else:
            raise ValueError(...)
    img_mem = sum(imgs_bytes)
    return img_mem
```

For `.pkg.slp` files, `label.image` triggers HDF5 decompression and read for each frame. This is done sequentially (no parallelism) over all 12K frames, taking ~21 minutes. The loaded images are then immediately discarded. The actual `_fill_cache()` call later re-reads all images again — this time with parallel workers (`cache_workers: 16`), which is why it's much faster (~9 min).

**Key issue**: Images are loaded from HDF5 twice — once for estimation, once for caching. The estimation path is sequential and has no progress logging, making it appear as a hang.

## Relevant dataloader/caching config

```yaml
data_pipeline_fw: torch_dataset_cache_img_memory
parallel_caching: true
cache_workers: 16

train_data_loader:
  num_workers: 4    # triggers worker memory overhead estimation
val_data_loader:
  num_workers: 4
```

The `num_workers: 4` setting is what triggers the worker memory overhead branch in `estimate_cache_memory()` (line 183-194), which requires knowing the raw cache size — hence the full image read.

## Suggested fixes

1. **Estimate from metadata instead of loading images**: Use the HDF5 dataset shape and dtype to compute `nbytes` without decompressing. For a `.pkg.slp` file, something like `video.shape[0] * np.prod(video.shape[1:]) * dtype.itemsize` or reading a single frame and multiplying.

2. **Sample-based estimation**: Read a small sample of images (e.g., 10-50 frames) and extrapolate to the full dataset.

3. **Combine estimation with caching**: If we're going to read all images anyway for caching, do the memory check on-the-fly during the first N frames and abort if memory is insufficient, rather than doing two full passes.

4. **At minimum, add progress logging** to `check_memory()` so it doesn't appear as a silent hang.

<details>
<summary>Full training config</summary>

```yaml
# Generalist centroid detector: 10K dataset, full config (scale=0.5)
# Usage: sleap-nn train configs/centroid_10k_full.yaml

data_config:
  train_labels_path:
    - datasets/generalist_centroid_10k/train.pkg.slp
  val_labels_path:
    - datasets/generalist_centroid_10k/val.pkg.slp
  test_file_path: datasets/generalist_centroid_10k/test.pkg.slp
  validation_fraction: 0.0
  user_instances_only: true
  data_pipeline_fw: torch_dataset_cache_img_memory
  parallel_caching: true
  cache_workers: 16

  preprocessing:
    ensure_rgb: false
    ensure_grayscale: true
    max_height: 512
    max_width: 640
    scale: 0.5

  use_augmentations_train: true
  augmentation_config:
    intensity:
      uniform_noise_min: 0.0
      uniform_noise_max: 0.04
      uniform_noise_p: 0.0
      gaussian_noise_mean: 0.0
      gaussian_noise_std: 0.02
      gaussian_noise_p: 0.0
      contrast_min: 0.8
      contrast_max: 1.2
      contrast_p: 0.0
      brightness_min: 0.9
      brightness_max: 1.1
      brightness_p: 0.0
    geometric:
      rotation_min: -180.0
      rotation_max: 180.0
      rotation_p: 1.0
      scale_min: 0.8
      scale_max: 1.2
      scale_p: 1.0
      translate_width: 0.0
      translate_height: 0.0
      translate_p: 0.0
      affine_p: 0.0
      erase_scale_min: 0.0001
      erase_scale_max: 0.01
      erase_ratio_min: 1.0
      erase_ratio_max: 1.0
      erase_p: 0.0
      mixup_lambda_min: 0.01
      mixup_lambda_max: 0.05
      mixup_p: 0.0

model_config:
  init_weights: default
  pretrained_backbone_weights: null
  pretrained_head_weights: null
  backbone_config:
    unet:
      in_channels: 1
      kernel_size: 3
      filters: 32
      filters_rate: 1.5
      max_stride: 32
      stem_stride: null
      middle_block: true
      up_interpolate: true
      stacks: 1
      convs_per_block: 2
      output_stride: 2
    convnext: null
    swint: null
  head_configs:
    single_instance: null
    centroid:
      confmaps:
        anchor_part: centroid
        sigma: 3.0
        output_stride: 2
    centered_instance: null
    bottomup: null
    multi_class_bottomup: null
    multi_class_topdown: null

trainer_config:
  train_data_loader:
    batch_size: 8
    num_workers: 4
    shuffle: true
  val_data_loader:
    batch_size: 8
    num_workers: 4
    shuffle: false

  model_ckpt:
    save_top_k: 1
    save_last: true

  max_epochs: 200
  min_train_steps_per_epoch: 200
  train_steps_per_epoch: null
  seed: 42

  trainer_devices: 1
  trainer_accelerator: auto
  trainer_strategy: auto

  save_ckpt: true
  ckpt_dir: models
  run_name: centroid-10k-full-animals
  resume_ckpt_path: null

  # --- WandB ---
  use_wandb: true
  wandb:
    entity: talmolab
    project: dreem-generalist-detector-animals
    name: centroid-10k-full-animals
    group: 10k
    wandb_mode: null
    api_key: null
    prv_runid: null
    current_run_id: null
    save_viz_imgs_wandb: true
    viz_enabled: true
    viz_boxes: false
    viz_masks: false
    viz_box_size: 5.0
    viz_confmap_threshold: 0.1
    log_viz_table: false
    delete_local_logs: null

  # --- Optimizer ---
  optimizer_name: Adam
  optimizer:
    lr: 0.0001
    amsgrad: false

  # --- LR scheduler ---
  lr_scheduler:
    step_lr: null
    reduce_lr_on_plateau:
      threshold: 0.01
      threshold_mode: rel
      cooldown: 5
      patience: 10
      factor: 0.5
      min_lr: 1.0e-08
    cosine_annealing_warmup: null
    linear_warmup_linear_decay: null

  # --- Early stopping ---
  early_stopping:
    stop_training_on_plateau: true
    patience: 20
    min_delta: 1.0e-08

  # --- Evaluation ---
  eval:
    enabled: true
    frequency: 1
    oks_stddev: 0.025
    oks_scale: null
    match_threshold: 50.0

  # --- Visualization ---
  visualize_preds_during_training: true
  keep_viz: false
  enable_progress_bar: false
```

</details>

## Environment

- sleap-nn 0.1.2
- Python 3.13.5
- PyTorch 2.10.0+cu130
- CUDA 13.0
- 2 GPUs available, 1 used for training
